google/cloud/internal/invoke_result.h

```diff
@@ -0,0 +1,180 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_INVOKE_RESULT_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_INVOKE_RESULT_H_
+
+#include "google/cloud/version.h"
+#include <type_traits>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+/**
+ * A helper to detect if expressions are valid and use SFINAE.
+ *
+ * This class will be introduced in C++17, it makes it easier to write SFINAE
+ * expressions, basically you use `void_t< some expression here >` in the SFINAE
+ * branch that is trying to test if `some expression here` is valid. If the
+ * expression is valid it expands to `void` and your branch continues, if it is
+ * invalid it fails and the default SFINAE branch is used.
+ *
+ * @see https://en.cppreference.com/w/cpp/types/void_t for more details about
+ *   this class.
+ */
+template <typename...>
+using void_t = void;
+
+/**
+ * A helper class to determine the result type of a function-like object.
+ *
+ * We want to define a class similar to `std::invoke_result<>` from C++17, to do
+ * so we need to know the result type of calling the function-like `F` with some
+ * parameters. But knowing how to call it depends on what `F` is:
+ *
+ * For function-like objects you would want to test the expression `f(a...)`,
+ * where `f` is of type `F` and `a...` are the arguments
+ *
+ * For pointer to member-like objects you would want to test the expression
+ * `a->f(b...)` where a is the first parameter (if a pointer) and `b...` are the
+ * remaining parameters.
+ *
+ * This of course gets more complicated with references (`a.f(b...)`), derived
+ * classes, etc.
+ *
+ * Fortunately, we just need to implement the easy case (function-like things)
+ * in the library:
+ *
+ * @tparam T the decayed (via `std::decay`) type of F.
+ */
+template <typename T>
```

Comment:
It doesnt appear that tparam T is used anywhere. Does it get referenced in the auto return type?

Reply:
You are correct. I added a comment to explain why.